libcst/matchers/_matcher_base.py

```diff
@@ -0,0 +1,638 @@
+                    nodes[1:],
+                    [AtMostN(matcher.matcher, n=matcher.n - 1), *matchers[1:]],
+                ):
+                    return True
+            # Finally, assume that this does not match the current node.
+            # Consume the matcher but not the node.
+            if _sequence_matches(nodes, matchers[1:]):
+                return True
+        elif isinstance(matcher, AtLeastN):
+            if matcher.n > 0:
+                # Only match if we can consume one of the matches, since we still
+                # need to match N nodes.
+                if _attribute_matches(nodes[0], matcher.matcher) and _sequence_matches(
+                    nodes[1:],
+                    [AtLeastN(matcher.matcher, n=matcher.n - 1), *matchers[1:]],
+                ):
+                    return True
+            else:
+                # First, assume that this does match a node (greedy).
+                # Consume one node since it matched this matcher.
+                if _attribute_matches(nodes[0], matcher.matcher) and _sequence_matches(
+                    nodes[1:], matchers
+                ):
+                    return True
+                # Now, assume that this does not match the current node.
+                # Consume the matcher but not the node.
+                if _sequence_matches(nodes, matchers[1:]):
+                    return True
+        else:
+            # There are no other types of wildcard consumers, but we're making
+            # pyre happy with that fact.
+            raise Exception(f"Logic error unrecognized wildcard {type(matcher)}!")
+    elif _matches(node, matcher):
+        # These values match directly
+        return _sequence_matches(nodes[1:], matchers[1:])
+
+    # Failed recursive case, no match
+    return False
+
+
+_AttributeValueT = Optional[Union[MaybeSentinel, libcst.CSTNode, str, bool]]
+_AttributeMatcherT = Optional[Union[BaseMatcherNode, DoNotCareSentinel, str, bool]]
+
+
+def _attribute_matches(  # noqa: C901
+    node: Union[_AttributeValueT, Sequence[_AttributeValueT]],
+    matcher: Union[_AttributeMatcherT, Sequence[_AttributeMatcherT]],
+) -> bool:
+    if isinstance(matcher, DoNotCareSentinel):
+        # We don't care what this is, so don't penalize a non-match.
+        return True
+    if isinstance(matcher, InverseOf):
+        # Return the opposite evaluation
+        return not _attribute_matches(node, matcher.matcher)
+
+    if isinstance(matcher, MatchIfTrue):
+        # We should only return if the matcher function is true.
+        return matcher.func(node)
+
+    if matcher is None:
+        # Should exactly be None
+        return node is None
+
+    if isinstance(matcher, str):
+        # Should exactly match matcher text
+        return node == matcher
+
+    if isinstance(matcher, bool):
+        # Should exactly match matcher bool
+        return node is matcher
```

Comment:
This looks like a base case worth a test case.
https://codecov.io/gh/Instagram/LibCST/pull/75/diff?src=pr&el=tree#D6-491